apisix/healthcheck_manager.lua

```diff
@@ -0,0 +1,313 @@
+            local upstream = res_conf.value.upstream or res_conf.value
+            local new_version = _M.upstream_version(res_conf.modifiedIndex, upstream._nodes_ver)
+            core.log.info("checking waiting pool for resource: ", resource_path,
+                    " current version: ", new_version, " requested version: ", resource_ver)
+            if resource_ver ~= new_version then
+                goto continue
+            end
+
+            -- if a checker exists then delete it before creating a new one
+            local existing_checker = working_pool[resource_path]
+            if existing_checker then
+                existing_checker.checker:delayed_clear(10)
+                existing_checker.checker:stop()
+                core.log.info("releasing existing checker: ", tostring(existing_checker.checker))
+            end
+            local checker = create_checker(upstream)
+            if not checker then
+                goto continue
+            end
+            core.log.info("create new checker: ", tostring(checker))
+            add_working_pool(resource_path, resource_ver, checker)
+        end
+
+        ::continue::
+        waiting_pool[resource_path] = nil
+    end
+end
+
+
+function _M.timer_working_pool_check()
+    if core.table.nkeys(working_pool) == 0 then
+        return
+    end
+
+    local working_snapshot = tab_clone(working_pool)
+    for resource_path, item in pairs(working_snapshot) do
+        --- remove from working pool if resource doesn't exist
+        local res_conf = fetch_latest_conf(resource_path)
+        if not res_conf or not res_conf.value then
+            item.checker:delayed_clear(DELAYED_CLEAR_TIMEOUT)
+            item.checker:stop()
+            core.log.info("try to release checker: ", tostring(item.checker))
+            working_pool[resource_path] = nil
+            goto continue
+        end
+        local current_ver = _M.upstream_version(res_conf.modifiedIndex,
+                                                res_conf.value._nodes_ver)
+        core.log.info("checking working pool for resource: ", resource_path,
+                    " current version: ", current_ver, " item version: ", item.version)
+        if item.version ~= current_ver then
+            item.checker:delayed_clear(DELAYED_CLEAR_TIMEOUT)
+            item.checker:stop()
+            core.log.info("try to release checker: ", tostring(item.checker))
+            working_pool[resource_path] = nil
+        end
+
+        ::continue::
+    end
+end
+
+function _M.init_worker()
+    local timer_create_checker_running = false
+    local timer_working_pool_check_running = false
+    timer_every(1, function ()
+        if not exiting() then
+            if timer_create_checker_running then
+                core.log.warn("timer_create_checker is already running, skipping this iteration")
+                return
+            end
+            timer_create_checker_running = true
+            pcall(_M.timer_create_checker)
```

Comment:
we always should capture the error message, and print them to error log

they are useful if the code run fail